### ISSUE TITLE
spring.cloud.stream.kafka.binder.brokers is ignored

### ISSUE THREAD

**Issue by NLatyshev** [spring-cloud/spring-cloud-stream-binder-kafka#1104]
Hi 
please look at      
https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/blob/13474bdafb82fb85d439c5a2a14e39c6992512dc/spring-cloud-stream-binder-kafka-core/src/main/java/org/springframework/cloud/stream/binder/kafka/properties/KafkaBinderConfigurationProperties.java#L407    
      
it ignores spring.cloud.stream.kafka.binder.brokers if spring.kafka.bootstrap-servers is set, is it desired behaviour?

**Comment by sobychacko** [comment]
The binder configuration (`spring.cloud.stream.kafka.binder.brokers`) if present should take precedence over the Boot one. (`spring.kafka.bootstrap-servers`). The code you pointed to is special logic in which the binder checks for the presence of Kafka's `bootstrap.servers` already present in the configuration (potentially set by Spring Boot). If this configuration is present and the binder configuration for bootstrap server differs from the default value (`localhost:9092`), then take the binder-provided value. The bottom line is that the binder-provided value always wins.

**Comment by NLatyshev** [comment]
cannot agree  with     

> If this configuration is present and the binder configuration for bootstrap server differs from the default value      
look at 
```
if (ObjectUtils.isEmpty(configuration.get(bootstrapServersConfig)) ||
				!kafkaConnectionString.equals("localhost:9092"))
```
if the config (`spring.kafka.bootstrap-servers`) is `NOT` present 'OR' the binder configuration for bootstrap server differs from the default value (`localhost:9092`), then take the binder-provided value      

I 've testesd it, but it was 3.0.6 and `spring.kafka.bootstrap-servers` has prcedence.  3.0.6 is very close to current

**Comment by sobychacko** [comment]
The first check almost always is `false` as Boot will set the default even if `spring.kafka.bootstrap-servers` is missing from the app configuration. So, the second checks if the binder configuration provided a different value. 

3.0.6 is an older release. 3.1.3 is the latest GA. In any case, if you are seeing issues, please provide a small reproducible sample where we can triage this issue further. Thanks!

**Comment by NLatyshev** [comment]
perhaps you are right about current version, because 3.0.6 
```
if (ObjectUtils.isEmpty(configuration.get(bootstrapServersConfig))) {
	configuration.put(bootstrapServersConfig, getKafkaConnectionString());
}
```
it was probably fixed

**Comment by sobychacko** [comment]
You are right. This was addressed only in the `3.0.9.RELEASE`.  See this [issue](https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/issues/967) and the associated [PR](https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/commit/112b95dada9d85f64e529c10040d595561475c92).

**Comment by sobychacko** [comment]
Closing the issue. Feel free to re-open if you see further issues.
